File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/Pathfinders/Flying/PathfinderGoalDiveOnTargetAttack.java
package com.offz.spigot.mobzy.Pathfinders.Flying;

import com.offz.spigot.mobzy.Mobs.Types.FlyingMob;
import net.minecraft.server.v1_13_R2.PathfinderGoal;
import org.bukkit.Location;
import org.bukkit.entity.LivingEntity;

public class PathfinderGoalDiveOnTargetAttack extends PathfinderGoal {
    private final FlyingMob mob;
    private final LivingEntity entity;
    private double diveVelocity;
    private Action currentAction = Action.FLY;
    private double minHeight;
    private double maxHeight;
    private double diveHeight;
    private double bashDuration = 30;
    private double bashLeft = bashDuration;
    private double bashVelX;
    private double bashVelZ;
    private double bashVelMultiplier = 0.6;
    private double startDiveHeightRange = 2;
    private double startDiveDistance = 8;

    public PathfinderGoalDiveOnTargetAttack(FlyingMob flying, double diveVelocity, double minHeight, double maxHeight, double startDiveDistance, double startDiveHeightRange, double bashVelMultiplier, double bashDuration) {
        this.mob = flying;
        this.diveVelocity = diveVelocity;
        this.minHeight = minHeight;
        this.maxHeight = maxHeight;
        this.startDiveDistance = startDiveDistance;
        this.startDiveHeightRange = startDiveHeightRange;
        this.bashVelMultiplier = bashVelMultiplier;
        this.bashDuration = bashDuration;
        diveHeight = pickDiveHeight();

        entity = ((LivingEntity) flying.getBukkitEntity());
    }

    public boolean a() {
        if (this.mob.getGoalTarget() == null) {
//            this.mob.yaw = -((float) MathHelper.c(this.mob.motX, this.mob.motZ)) * 57.295776F;
//            this.mob.aQ = this.mob.yaw;
            currentAction = Action.FLY;
            return false;
        }

        if (currentAction == Action.FLY)
            prepareDive();
        else if (currentAction == Action.DIVE)
            beginDive();
        else if (currentAction == Action.BASH)
            bash();
        return true;
    }

    public void prepareDive() {
        LivingEntity target = (LivingEntity) this.mob.getGoalTarget().getBukkitEntity();
        mob.lookAt(target);
        //if arrived to dive
        Location diveTarget = target.getLocation().clone().add(0, diveHeight, 0);
        if (mob.getLocation().distance(diveTarget) < startDiveDistance && Math.abs(mob.getLocation().getY() - diveTarget.getY()) < startDiveHeightRange) {
            diveHeight = pickDiveHeight();
            currentAction = Action.DIVE;
            return;
        }
        Location targetLoc = target.getLocation();
        mob.lookAt(target);

        mob.getControllerMove().a(targetLoc.getX(), targetLoc.getY() + diveHeight, targetLoc.getZ(), 1.0D);
    }

    public void beginDive() {
        LivingEntity target = (LivingEntity) this.mob.getGoalTarget().getBukkitEntity();
        mob.lookAt(target);
        Location targetLoc = target.getLocation();
        if (mob.distanceToEntity(target) < 2 || entity.getVelocity().getY() == 0 || mob.getY() < target.getLocation().getY() - 2) {
            currentAction = Action.BASH;
            bashVelX = entity.getLocation().getDirection().getX() * bashVelMultiplier;
            bashVelZ = entity.getLocation().getDirection().getZ() * bashVelMultiplier;
            return;
        }

        mob.getControllerMove().a(targetLoc.getX(), targetLoc.getY(), targetLoc.getZ(), 0.1D);
        entity.setVelocity(entity.getVelocity().setY(-Math.abs(diveVelocity)));

    }

    public void bash() {
        LivingEntity target = (LivingEntity) this.mob.getGoalTarget().getBukkitEntity();
        Location targetLoc = target.getLocation();
        if (bashLeft <= 0 || mob.distanceToEntity(target) < 2) {
            currentAction = Action.FLY;
            bashLeft = bashDuration;
            return;
        }

        mob.getControllerMove().a(targetLoc.getX(), mob.getY(), targetLoc.getZ(), 0.01D);

        entity.setVelocity(entity.getVelocity().setX(bashVelX).setZ(bashVelZ));
//        mob.lookAt(target);

        bashLeft--;
    }

    private double pickDiveHeight() {
        return Math.random() * (maxHeight - minHeight) + minHeight;
    }

    private enum Action {
        FLY, DIVE, BASH
    }
}

File: src/main/java/com/mineinabyss/mobzy/pathfinders/flying/PathfinderGoalDiveOnTargetAttack.kt
package com.mineinabyss.mobzy.pathfinders.flying

import com.mineinabyss.mobzy.mobs.types.FlyingMob
import com.mineinabyss.mobzy.pathfinders.MobzyPathfinderGoal
import kotlin.math.abs
import kotlin.random.Random

class PathfinderGoalDiveOnTargetAttack(
        override val mob: FlyingMob,
        private val diveVelocity: Double = -0.3,
        private val minHeight: Double = 6.0,
        private val maxHeight: Double = 10.0,
        private val startDiveDistance: Double = 16.0,
        private val startDiveHeightRange: Double = 2.0,
        private val bashVelMultiplier: Double = 0.6,
        private val bashDuration: Double = 30.0) : MobzyPathfinderGoal() {
    private var currentAction = Action.FLY
    private var diveHeight: Double = pickDiveHeight()
    private var bashLeft = bashDuration
    private var bashVelX = 0.0
    private var bashVelZ = 0.0

    override fun shouldExecute() = target != null

    override fun execute() = when (currentAction) {
        Action.FLY -> prepareDive()
        Action.DIVE -> beginDive()
        Action.BASH -> bash()
    }

    override fun shouldKeepExecuting() = target != null

    override fun reset() {
        bashLeft = bashDuration
        currentAction = Action.FLY
        moveController.a() //this should reset the controller's destination FIXME this method doesn't actually do that
    }

    private fun prepareDive() {
        val target = target ?: return
        mob.lookAtPitchLock(target)

        //if arrived to dive
        val diveTarget = target.location.clone().add(0.0, diveHeight, 0.0)
        if (diveTarget.distance(target.location) < startDiveDistance && abs(mob.locY - diveTarget.y) < startDiveHeightRange) {
            diveHeight = pickDiveHeight()
            currentAction = Action.DIVE
            return
        }

        val targetLoc = target.location
        moveController.a(targetLoc.x, targetLoc.y + diveHeight, targetLoc.z, 1.0) //TODO use controllerMove wrapper
    }

    private fun beginDive() {
        val target = target ?: return
        mob.lookAtPitchLock(target)
        val targetLoc = target.location
        if (mob.distanceTo(target) < 2 || entity.velocity.y == 0.0 || mob.locY <= target.location.y + 1.0) {
            currentAction = Action.BASH
            bashVelX = entity.location.direction.x * bashVelMultiplier
            bashVelZ = entity.location.direction.z * bashVelMultiplier
            return
        }
        moveController.a(targetLoc.x, targetLoc.y, targetLoc.z, 1.0)
        entity.velocity = entity.velocity.setY(-abs(diveVelocity))
    }

    private fun bash() {
        mob.lookAt(mob.locX + bashVelX, mob.locZ + bashVelZ)
        entity.velocity = entity.velocity.setX(bashVelX).setZ(bashVelZ)
        if (bashLeft-- <= 0 || target == null || mob.distanceTo(target!!) < 2 || entity.velocity.x == 0.0 || entity.velocity.z == 0.0) {
            currentAction = Action.FLY
            bashLeft = bashDuration
        }
    }

    private fun pickDiveHeight() = Random.nextDouble(minHeight, maxHeight)

    private enum class Action {
        FLY, DIVE, BASH
    }
}